truncate_client/src/lil_bits/battle.rs

```diff
@@ -17,69 +22,161 @@ impl<'a> BattleUI<'a> {
     }
 }
 
-fn render_word(battle_word: &BattleWord, ctx: &GameCtx, ui: &mut egui::Ui) {
-    let galley = ui.painter().layout_no_wrap(
-        battle_word.word.clone(),
+fn get_galleys<'a>(
+    battle_words: &'a Vec<BattleWord>,
+    ctx: &GameCtx,
+    ui: &mut egui::Ui,
+) -> Vec<Arc<Galley>> {
+    let dot = ui.painter().layout_no_wrap(
+        "•".into(),
         FontId::new(
-            ctx.theme.letter_size,
+            ctx.theme.letter_size * 0.75,
             egui::FontFamily::Name("Truncate-Heavy".into()),
         ),
-        match battle_word.valid {
-            Some(true) => ctx.theme.addition,
-            Some(false) => ctx.theme.defeated,
-            None => ctx.theme.outlines,
-        },
+        ctx.theme.outlines.darken(),
     );
-    let (rect, resp) = ui.allocate_at_least(galley.size(), Sense::hover());
-    ui.painter().galley(rect.min, galley);
 
-    resp.on_hover_ui(|ui| match (battle_word.valid, &battle_word.meanings) {
-        (None, _) => {
-            ui.label("Word did not need to be checked as the attack was invalid");
-        }
-        (Some(true), None) => {
-            ui.label("Valid word with no definition found");
-        }
-        (Some(true), Some(meanings)) => {
-            for meaning in meanings {
-                ui.label(format!("{}:", meaning.pos));
-                for def in &meaning.defs {
-                    ui.label(format!("  • {def}"));
-                }
+    let mut words: Vec<_> = battle_words
+        .iter()
+        .flat_map(|w| {
+            [
+                ui.painter().layout_no_wrap(
+                    w.word.clone(),
+                    FontId::new(
+                        ctx.theme.letter_size * 0.75,
+                        egui::FontFamily::Name("Truncate-Heavy".into()),
+                    ),
+                    match w.valid {
+                        Some(true) => ctx.theme.addition.darken(),
+                        Some(false) => ctx.theme.defeated.darken(),
+                        None => ctx.theme.outlines.darken().darken(),
+                    },
+                ),
+                dot.clone(),
+            ]
+        })
+        .collect();
+    words.pop();
+    words
+}
+
+fn paint_galleys<'a>(
+    mut galleys: Vec<Arc<Galley>>,
+    ui: &mut egui::Ui,
+    centered: bool,
+) -> egui::Response {
+    galleys.reverse();
```

Comment:
What's this for? Feels like it could be removed and then we'd just replace `last()` with `first()`.

Reply:
I should probably benchmark it, but I had it as a one-time performance hit of reversing the Vec so we can `pop()` off the end (see `let galley = galleys.pop().unwrap();` below). As opposed to removing the first item each iteration and shifting the elements to the left.

In hindsight it might be cheapest to avoid mutating the Vec at all and just clone the Arc